src/CommunityToolkit.Maui/ImageSources/GravatarImageSource.shared.cs

```diff
@@ -0,0 +1,239 @@
+	}
+
+	/// <summary>Returns the Uri as a string.</summary>
+	/// <returns>String of the URI.</returns>
+	public override string ToString()
+	{
+		return $"Uri: {Uri}, Email:{Email}, Size:{GravatarSize}, Image: {DefaultGravatarName(Image)}, CacheValidity:{CacheValidity}, CachingEnabled:{CachingEnabled}";
+	}
+
+	/// <summary>On parent set.</summary>
+	protected override void OnParentSet()
+	{
+		object element = Parent;
+		base.OnParentSet();
+		if (element is not VisualElement parentElement || parentElement is null)
+		{
+			GravatarSize = defaultSize;
+			return;
+		}
+
+		SetBinding(ParentWidthProperty, new Binding(nameof(VisualElement.Width), BindingMode.OneWay, source: parentElement));
+		SetBinding(ParentHeightProperty, new Binding(nameof(VisualElement.Height), BindingMode.OneWay, source: parentElement));
+	}
+
+	static string DefaultGravatarName(DefaultImage defaultGravatar)
+			=> defaultGravatar switch
+			{
+				DefaultImage.FileNotFound => "404",
+				DefaultImage.MysteryPerson => "mp",
+				_ => $"{defaultGravatar}".ToLower(),
+			};
+
+	static string GetMd5Hash(ReadOnlySpan<char> str)
+	{
+		using var md5 = MD5.Create();
+		Span<byte> hash = md5.ComputeHash(Encoding.UTF8.GetBytes(str.ToArray()));
+		return BitConverter.ToString(hash.ToArray(), 0, hash.Length).Replace("-", string.Empty, StringComparison.OrdinalIgnoreCase).ToLowerInvariant();
+	}
+
+	static void OnDefaultImagePropertyChanged(BindableObject bindable, object oldValue, object newValue)
+	{
+		GravatarImageSource gravatarImageSource = (GravatarImageSource)bindable;
+		gravatarImageSource.HandleNewUriRequested(gravatarImageSource.Email, (DefaultImage)newValue);
+	}
+
+	static void OnEmailPropertyChanged(BindableObject bindable, object oldValue, object newValue)
+	{
+		GravatarImageSource gravatarImageSource = (GravatarImageSource)bindable;
+		gravatarImageSource.HandleNewUriRequested((string?)newValue, gravatarImageSource.Image);
+	}
+
+	static void OnSizePropertyChanged(BindableObject bindable, object oldValue, object newValue)
+	{
+		GravatarImageSource gravatarImageSource = (GravatarImageSource)bindable;
+		if (newValue is not int intNewValue || intNewValue <= -1)
+		{
+			return;
+		}
+
+		if (gravatarImageSource.GravatarSize == 0)
+		{
+			gravatarImageSource.GravatarSize = intNewValue;
+			return;
+		}
+
+		gravatarImageSource.GravatarSize = Math.Min(intNewValue, gravatarImageSource.GravatarSize);
+		gravatarImageSource.HandleNewUriRequested(gravatarImageSource.Email, gravatarImageSource.Image);
+	}
+
+	async Task<Stream> DownloadStreamAsync(Uri? uri, CancellationToken cancellationToken)
+	{
+		try
+		{
+			return await StreamWrapper.GetStreamAsync(uri, cancellationToken, singletonHttpClient).ConfigureAwait(false);
+		}
+		catch (Exception ex)
+		{
+			Debug.WriteLine($"Error getting stream for {Uri}: {ex}");
+			return System.IO.Stream.Null;
+		}
+	}
+
+	void HandleNewUriRequested(string? email, DefaultImage image)
+	{
+		if (GravatarSize == 0)
```

Comment:
is it possible if GravatarSize is in range from 1 to 2048?

Reply:
The valid range to be set is between 1 and 2048 but the default is 0 to know it hasn't been set.

Reply:
It should have its default value. 0 seems like invalid value. I suggest to have default value equals let’s say 80. And remove this check. 0 is not a valid size

Reply:
If we use the default of 80 then we result in multiple HTTP calls which is inefficient

Reply:
But it's default value. Why does it call it multiple times? And also what is the difference between defaultSize and gravatarSize?

Reply:
I think a value of -1 is better than 0, as that's regarded as "un set".